### DISCUSSION TITLE
v0.8.0

### DISCUSSION THREAD

**Discussion by RobertCraigie** [RobertCraigie/prisma-client-py#663] [General]
## ⚠️ Prisma Schema Breaking Changes

This release contains some changes to the format of the Prisma Schema.

Most of these changes are due to a conceptual shift from allowing implicit behaviour to forcing verboseness to reduce the amount of under the hood magic that Prisma does, thankfully this means that a lot of the changes that you will be required to make should be pretty straightforward and easily fixable by running `prisma format` which will show you all the places that need changed in your schema.

Changes:
- [Explicit unique constraints for 1:1 relations](https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-4#explicit-unique-constraints-on-one-to-one-relations)
- [Removed support for usage of references on implicit m:n relations](https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-4#disallow-references-syntax-for-implicit-many-to-many-relations)
- [Enforcing uniqueness of referenced fields in the references argument in 1:1 and 1:m relations for MySQL](https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-4#enforced-use-of-unique-or-id-attribute-for-one-to-one-and-one-to-many-relations-mysql-and-mongodb)
- Removal of the `sqlite://` URL prefix, you should now use `file://` instead
- [Improved grammar for string literals](https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-4#enforced-use-of-unique-or-id-attribute-for-one-to-one-and-one-to-many-relations-mysql-and-mongodb)

For more details see the [Prisma v4.0.0 upgrade path](https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-4).

## ⚠️ Internal Raw Query Changes

This release includes an internal restructuring of how raw queries are deserialized. While all these changes should be completely backwards compatible, there may be edge cases that have changed. If you encounter one of these edge cases please [open an issue](https://github.com/RobertCraigie/prisma-client-py/issues/new/choose) and it will be fixed ASAP.

For some additional context, this restructuring means that most fields will internally be returned as strings until Prisma Client Python deserializes them (previously this was done at the query engine level).

## CLI Improvements

This release completely refactors how the Prisma CLI is downloaded and ran. The previous implementation relied on downloading a single [pkg](https://github.com/vercel/pkg) binary, this worked but had several limitations which means we now:

- Support ARM architectures & certain Linux distributions
- Support running Prisma Studio from the CLI
- Support generating the JS Client
- No longer have to maintain a separate implementation from Prisma for detecting the current platform & logic for downloading engines
  - As a user this means that more platforms & architectures will be supported faster!

The new solution is involves directly downloading a Node.js binary (if you don't already have it installed) and directly installing the Prisma ClI through `npm`.  Note that this does not pollute your userspace and does **not** make Node available to the rest of your system.

This will result in a small size increase (~150MB) in the case where Node is not already installed on your machine, if this matters to you you can install Prisma Client Python with the `node` extra, e.g. `pip install prisma[node]`,  which will install a Node binary to your `site-packages` that results in the same storage requirements as the previous `pkg` solution. You can also directly install [nodejs-bin](https://pypi.org/project/nodejs-bin/) yourself. It's also worth noting that this release includes significant (~50%) reduction in the size of the Prisma Engine binaries which makes the default Node binary size increase less impactful.

## Prisma Studio

With this release you can now run [Prisma Studio](https://www.prisma.io/studio) from the CLI which makes it incredibly easy to view & edit the data in your database. Simply run the following command

```
$ prisma studio
```

Or

```
$ prisma studio --schema=backend/schema.prisma
```

![Screenshow showcasing Prisma Studio usage](https://website-v8.vercel.app/images/studio-ui.png)

> Note that there is also a dark mode available

## Support for CockroachDB

This release adds official support for CockroachDB. You could've used CockroachDB previously by setting `provider` to `postgresql` but now you can explicitly specify CockroachDB in your Prisma Schema:

```prisma
datasource db {
  provider = "cockroachdb"
  url      = env("COCKROACHDB_URL")
}
```

It should be noted that there are a couple of edge cases:

- [BigInt ID fields cannot be atomically updated](https://github.com/prisma/prisma/issues/16612)
- [BigInt & Int fields cannot be atomically divided](https://github.com/prisma/prisma/issues/16511)
- [Array push operation is not supported](https://github.com/prisma/prisma/issues/13892)

## Prisma Updates

TL;DR for improvements made by Prisma that will now be in Prisma Client Python

- [[3.14.0] PostgreSQL extended indexes support (now generally available)](https://www.prisma.io/docs/concepts/components/prisma-schema/indexes#configuring-the-access-type-of-indexes-with-type-postgresql)
- [[4.0.0] Defaults for scalar lists](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#define-a-scalar-list-with-a-default-value)
- [[4.3.0] Prisma CLI exit code fixes](https://github.com/prisma/prisma/releases/tag/4.3.0#:~:text=Prisma%20CLI%20exit%20code%20fixes)
- [[4.5.0] PostgreSQL extension management](https://www.prisma.io/docs/concepts/components/prisma-schema/postgresql-extensions)
- [[4.7.0] Prisma Schema relationMode is now generally available](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/relation-mode)
- [[4.7.0] Query across multiple database schemas in PostgreSQL](https://www.prisma.io/docs/guides/database/multi-schema)
- [4.8.0] Decreased the size of the Query Engine binary ~50%
- [[4.8.0] Improved support for OpenSSL 3.x](https://github.com/prisma/prisma/issues/16553#issuecomment-1353302617)

Full list of changes:
- https://github.com/prisma/prisma/releases/tag/3.14.0
- https://github.com/prisma/prisma/releases/tag/3.15.0
- https://github.com/prisma/prisma/releases/tag/3.15.1
- https://github.com/prisma/prisma/releases/tag/3.15.2
- https://github.com/prisma/prisma/releases/tag/4.0.0
- https://github.com/prisma/prisma/releases/tag/4.1.0
- https://github.com/prisma/prisma/releases/tag/4.1.1
- https://github.com/prisma/prisma/releases/tag/4.2.0
- https://github.com/prisma/prisma/releases/tag/4.2.1
- https://github.com/prisma/prisma/releases/tag/4.3.0
- https://github.com/prisma/prisma/releases/tag/4.3.1
- https://github.com/prisma/prisma/releases/tag/4.4.0
- https://github.com/prisma/prisma/releases/tag/4.5.0
- https://github.com/prisma/prisma/releases/tag/4.6.0
- https://github.com/prisma/prisma/releases/tag/4.7.0
- https://github.com/prisma/prisma/releases/tag/4.7.1
- https://github.com/prisma/prisma/releases/tag/4.8.0

## Miscellaneous Changes

- [Added tests for MariaDB](https://github.com/RobertCraigie/prisma-client-py/pull/582)
- [Add validation to ensure model names do not clash with reserved keywords](https://github.com/RobertCraigie/prisma-client-py/issues/594)
- [Removed dev dependencies from the package distribution](https://github.com/RobertCraigie/prisma-client-py/pull/610)
- [Fixed incorrect type references under certain conditions for partial models](https://github.com/RobertCraigie/prisma-client-py/pull/646)
- [Added tests to ensure unsupported features for a given database do not pass type checks](https://github.com/RobertCraigie/prisma-client-py/pull/653)
- [Fixed docs typo](https://github.com/RobertCraigie/prisma-client-py/pull/656) - thanks @HigherOrderLogic!

## Public Roadmap

Going forward we will now use a [GitHub Project](https://github.com/users/RobertCraigie/projects/2/views/1) to track state and relative priority of certain issues. If you'd like to increase the priority of issues that would benefit you please add 👍 reactions.

This is less of a roadmap per se but will hopefully give you some insight into the priority of given issues / features.

## Attributions

Thank you to @kfields for helping with raw query deserialization!

Massive thank you to @prisma & @techied for their continued support and @exponential-sponsorship for becoming a sponsor!

![sponsors](https://user-images.githubusercontent.com/23125036/210181025-93808df5-a9c3-410e-9e9d-a29ace34b464.png)

<hr /><em>This discussion was created from the release <a href='https://github.com/RobertCraigie/prisma-client-py/releases/tag/v0.8.0'>v0.8.0</a>.</em>

**Comment by ezorita** [comment]
Awesome release @RobertCraigie!

How would one now trigger the binary download? In the previous version, running `prisma generate` without further arguments would trigger the binary download and then fail, but it does not seem to be the case in the current version.
